### ISSUE TITLE
Provide a cluster worker pool

### ISSUE THREAD

**Issue by Shinigami92** [pioardi/poolifier#66]
I have rewritten your `FixedThreadPool` to a `FixedClusterPool` :sparkles:  
When my TypeScript branch (#64) is merged, I will create a new PR to also let poolifier provide a `Process/Cluster Pool` 

If you want, you could give me repo write permissions with the possibility to only create branches but not access or merge into master branch
Then we can work together on this PR together (e.g. write tests)

**Comment by Shinigami92** [comment]
My current code:
(_comments and some variable names need to be changed_)

<details>
<summary>fixed.ts</summary>

```ts
// Christopher will create a PR in poolifier repo
// Therefore this code is temporarily

import { SendHandle } from 'child_process';
import * as cluster from 'cluster';
import { Worker } from 'cluster';

function empty(): void {
    // empty
}

export type Draft<T> = { -readonly [P in keyof T]?: T[P] };

export type WorkerWithMessageChannel = Worker; // & Draft<MessageChannel>;

export interface FixedClusterPoolOptions {
    /**
     * A function that will listen for error event on each worker thread.
     */
    errorHandler?: (this: Worker, e: Error) => void;
    /**
     * A function that will listen for online event on each worker thread.
     */
    onlineHandler?: (this: Worker) => void;
    /**
     * A function that will listen for exit event on each worker thread.
     */
    exitHandler?: (this: Worker, code: number) => void;
    /**
     * This is just to avoid not useful warnings message, is used to set `maxListeners` on event emitters (workers are event emitters).
     *
     * @default 1000
     */
    maxTasks?: number;
}

/**
 * A thread pool with a static number of threads, is possible to execute tasks in sync or async mode as you prefer.
 *
 * This pool will select the worker thread in a round robin fashion.
 *
 * @author [Alessandro Pio Ardizio](https://github.com/pioardi)
 * @since 0.0.1
 */
export class FixedClusterPool<Data = any, Response = any> {
    public readonly workers: WorkerWithMessageChannel[] = [];
    public nextWorker: number = 0;

    // threadId as key and an integer value
    public readonly tasks: Map<WorkerWithMessageChannel, number> = new Map<WorkerWithMessageChannel, number>();

    protected _id: number = 0;

    /**
     * @param numThreads Num of threads for this worker pool.
     * @param filePath A file path with implementation of `ThreadWorker` class, relative path is fine.
     * @param opts An object with possible options for example `errorHandler`, `onlineHandler`. Default: `{ maxTasks: 1000 }`
     */
    public constructor(
        public readonly numThreads: number,
        public readonly filePath: string,
        public readonly opts: FixedClusterPoolOptions = { maxTasks: 1000 }
    ) {
        if (!cluster.isMaster) throw new Error('Cannot start a cluster pool from a worker!');
        if (!this.filePath) throw new Error('Please specify a file with a worker implementation');

        cluster.setupMaster({
            exec: this.filePath,
        });

        for (let i: number = 1; i <= this.numThreads; i++) {
            this._newWorker();
        }
    }

    public destroy(): void {
        for (const worker of this.workers) {
            worker.kill();
        }
    }

    /**
     * Execute the task specified into the constructor with the data parameter.
     *
     * @param data The input for the task specified.
     * @returns Promise that is resolved when the task is done.
     */
    public async execute(data: Data): Promise<Response> {
        // configure worker to handle message with the specified task
        const worker: WorkerWithMessageChannel = this._chooseWorker();
        // console.log('FixedClusterPool#execute choosen worker:', worker);
        this.tasks.set(worker, this.tasks.get(worker) + 1);
        const id: number = ++this._id;
        const res: Promise<Response> = this._execute(worker, id);
        // console.log('FixedClusterPool#execute send data to worker:', worker);
        worker.send({ data: data || {}, _id: id });
        return res;
    }

    protected _execute(worker: WorkerWithMessageChannel, id: number): Promise<Response> {
        return new Promise((resolve, reject) => {
            const listener: (message: any, handle: SendHandle) => void = (message) => {
                // console.log('FixedClusterPool#_execute listener:', message);
                if (message._id === id) {
                    worker.removeListener('message', listener);
                    this.tasks.set(worker, this.tasks.get(worker) - 1);
                    if (message.error) reject(message.error);
                    else resolve(message.data);
                }
            };
            worker.on('message', listener);
        });
    }

    protected _chooseWorker(): WorkerWithMessageChannel {
        if (this.workers.length - 1 === this.nextWorker) {
            this.nextWorker = 0;
            return this.workers[this.nextWorker];
        } else {
            this.nextWorker++;
            return this.workers[this.nextWorker];
        }
    }

    protected _newWorker(): WorkerWithMessageChannel {
        const worker: WorkerWithMessageChannel = cluster.fork();
        worker.on('error', this.opts.errorHandler || empty);
        worker.on('online', this.opts.onlineHandler || empty);
        // TODO handle properly when a thread exit
        worker.on('exit', this.opts.exitHandler || empty);
        this.workers.push(worker);
        // const { port1, port2 } = new MessageChannel();
        // worker.send({ parent: port1 }, [port1]);
        // worker.port1 = port1;
        // worker.port2 = port2;
        // we will attach a listener for every task,
        // when task is completed the listener will be removed but to avoid warnings we are increasing the max listeners size
        // worker.port2.setMaxListeners(this.opts.maxTasks || 1000);
        worker.setMaxListeners(this.opts.maxTasks || 1000);
        // init tasks map
        this.tasks.set(worker, 0);
        return worker;
    }
}
```

</details>

<details>
<summary>workers.ts</summary>

```ts
// Christopher will create a PR in poolifier repo
// Therefore this code is temporarily

import { AsyncResource } from 'async_hooks';
import * as cluster from 'cluster';

export interface ClusterWorkerOptions {
    /**
     * Max time to wait tasks to work on (in ms), after this period the new worker threads will die.
     *
     * @default 60.000 ms
     */
    maxInactiveTime?: number;
    /**
     * `true` if your function contains async pieces, else `false`.
     *
     * @default false
     */
    async?: boolean;
}

type MessageValue<Data> = { readonly data?: Data; readonly _id?: number; readonly kill?: number };

/**
 * An example worker that will be always alive, you just need to **extend** this class if you want a static pool.
 *
 * When this worker is inactive for more than 1 minute, it will send this info to the main thread,
 * if you are using DynamicThreadPool, the workers created after will be killed, the min num of thread will be guaranteed.
 *
 * @author [Alessandro Pio Ardizio](https://github.com/pioardi)
 * @since 0.0.1
 */
export class ClusterWorker<Data = any, Response = any> extends AsyncResource {
    protected readonly maxInactiveTime: number;
    protected readonly async: boolean;
    protected lastTask: number;
    protected readonly interval: NodeJS.Timeout;
    // protected parent: MessagePort;

    public constructor(fn: (data: Data) => Response, public readonly opts: ClusterWorkerOptions = {}) {
        super('worker-cluster-pool:pioardi');

        this.maxInactiveTime = this.opts.maxInactiveTime || 1000 * 60;
        this.async = !!this.opts.async;
        this.lastTask = Date.now();
        if (!fn) throw new Error('Fn parameter is mandatory');
        // keep the worker active
        if (!cluster.isMaster) {
            // console.log('ClusterWorker#constructor', 'is not master');
            this.interval = setInterval(this._checkAlive.bind(this), this.maxInactiveTime / 2);
            this._checkAlive.bind(this)();
        }
        cluster.worker.on('message', (value: MessageValue<Data>) => {
            // console.log("cluster.on('message', value)", value);
            if (value && value.data && value._id) {
                // here you will receive messages
                // console.log('This is the main thread ' + isMainThread)
                if (this.async) {
                    this.runInAsyncScope(this._runAsync.bind(this), this, fn, value);
                } else {
                    this.runInAsyncScope(this._run.bind(this), this, fn, value);
                }
                // } else if (value.parent) {
                //     // save the port to communicate with the main thread
                //     // this will be received once
                //     this.parent = value.parent;
            } else if (value.kill) {
                // here is time to kill this thread, just clearing the interval
                clearInterval(this.interval);
                this.emitDestroy();
            }
        });
    }

    protected _checkAlive(): void {
        if (Date.now() - this.lastTask > this.maxInactiveTime) {
            cluster.worker.send({ kill: 1 });
        }
    }

    protected _run(fn: (data: Data) => Response, value: MessageValue<Data>): void {
        try {
            const res: Response = fn(value.data);
            cluster.worker.send({ data: res, _id: value._id });
            this.lastTask = Date.now();
        } catch (e) {
            cluster.worker.send({ error: e, _id: value._id });
            this.lastTask = Date.now();
        }
    }

    protected _runAsync(fn: (data: Data) => Promise<Response>, value: MessageValue<Data>): void {
        fn(value.data)
            .then((res) => {
                cluster.worker.send({ data: res, _id: value._id });
                this.lastTask = Date.now();
            })
            .catch((e) => {
                cluster.worker.send({ error: e, _id: value._id });
                this.lastTask = Date.now();
            });
    }
}
```

</details>

I need to build `DynamicClusterPool` later on

**Comment by jerome-benoit** [comment]
You can stack your PRs and create dependency between, that will ease the code review.

**Comment by Shinigami92** [comment]
I'm not a friend of stacking branches on top of each other, because this makes branch management really hard if there is a conflict or commit in between that suddenly leads into a new direction

I have waited for your response 2+ month :smile: ... so I think I have the time and the calm to just wait a little bit more and make one step after each other :slightly_smiling_face:

**Comment by jerome-benoit** [comment]
> I'm not a friend of stacking branches on top of each other, because this makes branch management really hard if there is a conflict or commit in between that suddenly leads into a new direction

You're right. That's why I've always considered the one PR per feature a mistake. The commits level is way more appropriate and in fact used by most of the prominent FOSS project like Linux Kernel, git, etc.

> I have waited for your response 2+ month 😄 ... so I think I have the time and the calm to just wait a little bit more and make one step after each other 🙂

Just to be clear: I'm not an official maintainer of poolifier but I'm very interested by your cluster worker pool integration. So I'm making what I think is necessary to speed up the merge of your work ;)

In any case, the final decision stay in @pioardi hands.

**Comment by Shinigami92** [comment]
Oh wow I was so confused about the name, I just thought (until now) that you just renamed your account :rofl:  
But you are a whole different person LOL

**Comment by jerome-benoit** [comment]
The goal of my reviews are just to: 
+ offload some review work from the official maintainer; 
+ make hopefully some useful suggestions that will ease your work acceptance;
+ shows that your work interests some poolifier users. 

Let's hope that it will speeds up your work merge.

**Comment by jerome-benoit** [comment]
> I'm not a friend of stacking branches on top of each other, because this makes branch management really hard if there is a conflict or commit in between that suddenly leads into a new direction

But still, could you please publish your work in progress regarding cluster worker pool integration in a branch under your poolifier fork? 
I would like to test it.

**Comment by pioardi** [comment]
Thanks both @jerome-benoit  and @Shinigami92 .
 My main focus is to listen what the "community" ask about this library to move on in the right direction so thanks to both of you.
As you know my preference is to promote backward compatibility where possible. 

Said that I will need in the near future to have repo mainteners that help me with this library and I am considering both of you for that, we could probably need to have a chat or something like that, everything at its time :)

**Comment by Shinigami92** [comment]
> But still, could you please publish your work in progress regarding cluster worker pool integration in a branch under your poolifier fork?
> I would like to test it.

Yeah I need to set it up, maybe not today anymore, give me some days or so, because I think I don't have it right now in a branch as you think. That's why I copied it into the [comment](https://github.com/pioardi/poolifier/issues/66#issuecomment-726111277) 2 month ago


> [...] mainteners that help me with this library [...]

I'm happy to help and become a (side-/co-)maintainer  
I have some other projects and communities, but I think this lib is such small that I also can help and have the time for that :slightly_smiling_face:

**Comment by jerome-benoit** [comment]
Is Worker from cluster module API equivalent to worker_threads? 
If yes, code reuse can probably be done easily by using some TS magic regarding typing.

**Comment by Shinigami92** [comment]
@jerome-benoit I need to refresh my 2-month-old-mind when working on this (I hope so today)  
But first I would prefer to make the base project and structure stable so this PR produces as less as  possible conflicts
They more branches I need to maintain in parallel they more I need to manage the state of each of the branches
If I can work step by step, I'm more efficient

**Comment by Shinigami92** [comment]
NOW the time has come :tada: 
@jerome-benoit @pioardi I will tackle this now!

I'm also available via gitter the whole day today, because now it's clear that I have a so called slack-day
A day in each month where I can spend my working time for what ever I want :slightly_smiling_face:

**Comment by jerome-benoit** [comment]
We maybe need to precise that the cluster worker pool will only work with stateless network protocols. 
By default cluster do not offer a sticky load balancing anyway, but with a pool, it will be not easy to add a session affinity mechanism (let's say with HTTP first) to one of the worker in the pool.
